SU2_CFD/include/drivers/CDiscAdjMultizoneDriver.hpp

```diff
@@ -79,6 +79,10 @@ class CDiscAdjMultizoneDriver : public CMultizoneDriver {
   unsigned short* direct_nInst;   /*!< \brief Total number of instances in the direct problem. */
   unsigned short* nInnerIter;     /*!< \brief Number of inner iterations for each zone. */
 
+  /*!< \brief Individual cross-terms of the coupled problem, 5D array [iZone][jZone][iSol](iPoint,iVar).
+              The column sum, i.e. along all iZones for each jZone, gives the External (total cross-term)
+              for jZone, we need to store all terms to have BGS-type updates with relaxation. */
+  vector<vector<vector<su2passivematrix> > > Cross_Terms;
```

Comment:
@oleburghardt, needs cleanup but I'll go with your suggestion for a nZone*nZone structure.

Reply:
@oleburghardt I think I've got it working. Thanks for pointing to the tutorial, now I know why all my values were close even when the code was wrong (10000 iterations is 3 orders of magnitude more than I am used to with FSI cases).

Just to summarize what I did, each cross term is stored separately in that `Cross_Terms` 5D array.
When a cross term is extracted, it is updated with relaxation (which we need in FSI but the default is no relaxation so no impact on CHT):
`x = alpha * x_new + (1-alpha) * x_old`
The delta in that update (`x-x_old`) is added to `External`, which carries the sum of all cross terms and the objective function gradient.
The latter is now only computed once, outside the outer iteration loop, and it initializes `External` (`External_Old` is not longer required).
In the inner iteration loop `External` is added to `Solution` before computing the adjoints instead of after, this means when exiting the inner loop `Solution` does not contain cross terms or the OF gradient.
`Solution_BGS` and `Solution_BGS_k` also do not contain `External` therefore by restoring `Solution_BGS_k` to `Solution` before the inner iterations, and adding `External` to `Solution` as the first thing we do in the inner loop, we pickup exactly from where we stopped before (that is the intention anyway, the whole thing might be wrong again).
The cross terms are stored in `passivedouble` as that occupies half the space, we can think of streamlining things further by not using BGS_k, which is not needed since BGS == Solution.

I will compare the converge rate of this type of update with the one you had before, if it is worse I will put in the option to run as before (that may require the OF term to be carried separately).

Reply:
Conclusions:
- No negative impact in CHT convergence rate, after the 10k iterations the residuals are nearly the same and the gradients (running everything on 1 core) are -0.719645 for develop and -0.719302 for this branch.
- No negative impact in correctness, converging the CHT case to below 10^-8 the gradient with this branch is -0.717685 whether the fluid is evaluated first or third. 
- Within this framework FSI adjoints seem not to need relaxation, which is great!

Reply:
Thanks for the summary. Just looked through the code and saw that is should all be correct.
But just to be on the safe side I repeated all the tests (fluid zone at different places, you remember) and they all give the gradient above.